Inline citations

sondregronas opened the pull request:
Extracts affixes and citekeys in their own lists from markdown with a new regex. Also some error handling when citekeys are incorrect (i.e. when creating links for twitter - `[@twitter]` is now possible).

Could address #103 - seems to be working fine for now.

I've commented out `cite_prefix` and `cite_suffix` for now, not sure how to implement them so pandoc parses them correctly.

*NOTE*: Using affixes does break inline citation. Only proper `[@citekeys]` will be rendered as `[^n]`. `[see @johndoe, p155]` will be represented as `@johndoe` in `cite_keys` and in references (if applicable), but still render as `[see @johndoe, p155]` inline.

However I presume this would work in conjunction with what you've mentioned in #117 and potentially fix #77 - but I've got no time to verify or test it myself :)

sondregronas commented:
The tests look a bit messy now due to `find_cite_keys` returning a tuple, for now this seems to be working.

I've also added the prefix and suffix to pairs in `format_citations`

All inline citations from my testing render as expected, including those with suffixes/prefixes. Affixes are ignored for now, not sure how to pass them correctly to `format_simple|pandoc`, nor understand what the bibtex syntax would be - but this feels either close - or too complicated & way off :)

The only limitation I can see from my testing is that citekeys NEED to be separated by `; ` - though not sure what is common practice. I've seen some examples that go `[see @doe, p. 15-26 and @john, p. 33-37]` and I've got no clue what the best way to approach that would be right now.

sondregronas commented:
I got it working somewhat afterall - with some limitations. I reckon theres a way better way of doing this, but this "works" for now.

shyamd commented:
Really appreciate the effort here. The improved regex is great! I'm not a big fan of the affixes, since it's basically *another* custom citation language. The point about the `bibentry` method in #117  is that its a way to get the cite_key and citation separately via pandoc. That way anytime someone wants something more custom then the most basic basic style in general, they use pandoc and their choice of CSL styles. This prevents this plugin from supporting *another* custom citation language. 

I'll dig in this more this coming weekend.

sondregronas commented:
Is there any documentation around the `bibentry` syntax? I couldn't find any.

I've slept on it and removed the affixes from `find_cite_keys` since they can just be extracted by doing a `fullcite.split(key)` if they are needed - should make things easier to work with again, leaving only `fullcite` and `key(s)` in `pairs` which I presume is all thats necessary.

sondregronas commented:
I've got inline citations formatted correctly with this:

```python
def _convert_pandoc_citekey(bibtex_string, csl_path, fullcite):
  fullcite = '[see @test, p. 20]' # fullcite
  # using test.bib and harvard.csl
  with tempfile.TemporaryDirectory() as tmpdir:
      bib_path = Path(tmpdir).joinpath("temp.bib")
      with open(bib_path, "w") as bibfile:
          bibfile.write(bibtex_string)
  
      markdown = pypandoc.convert_text(
          source=fullcite,
          to="markdown-citations",
          format="markdown",
          extra_args=[
              "--citeproc",
              "--csl",
              csl_path,
              "--bibliography",
              bib_path,
          ],
      )
  return markdown.split("\r\n")[0].split("\n")[0]
``` 
This outputs `(see Author and Author, 2019, p. 20)` with `harvard-cite-them-right.csl`. `nature.csl` does not render correctly.

It DOES generate a bibliography entry too, but for some reason it omits the page numbers - maybe that's expected behavior?

Should be relatively easy to add an option to (optionally) add a `[^n]` to the end aswell, which I think might be sufficient?
Couldn't find a nice way to pass the `bibtex_string` to `insert_citation_keys`, but it would look something like this:
```python
replacement_citaton = "".join(["[^{}]".format(quad[2]) for quad in quad_group])
if use_inline:
    inline_citation = _convert_pandoc_citekey(bibtex_string, csl_file, full_citation)
    replacement_citaton = f"{inline_citation}{replacement_citaton}"
```

sondregronas commented:
I've got inline citations working now I think - not 100% certain of what the expected behavior is for the CSL file I tested with. Footnote links are still applied after the citation, this can be optionally toggled I suppose.

With `harvard.csl` the following: `Text [see @Bivort2016 and @test]` gets rendered as `Text Author and Author (2019)[^1]` and I'm not sure why that is.

`Text [see @Bivort2016; @test]` gets rendered as `Text (see De Bivort and Van Swinderen, 2016; Author and Author, 2019)[^1][^2]`

Will hopefully find some time to write tests later - `_convert_pandoc_new` should probably also be converted to use the full citation block aswell, which would simplify more things :)

shyamd commented:
A lot of things are happening here that makes it a bit difficult to review this PR. Correct me if my summary is wrong:
- Updates the cite_block parsing regex to be more resilient to badly formatted citations
- Fixes dropping of citations that are not in the bibtex file
- Adds inline citation rendering, IE Not just `[^1]` but `(Some body et al. 2019)[^1]`.

Is that all correct?

sondregronas commented:
That is indeed correct

shyamd commented on src/mkdocs_bibtex/utils.py:
```diff
@@ -145,8 +214,29 @@ def insert_citation_keys(citation_quads, markdown):
+                pandoc_version = pypandoc.get_pandoc_version()
+                pandoc_version_tuple = tuple(int(ver) for ver in pandoc_version.split("."))
+                if pandoc_version_tuple <= (2, 11):
+                    raise Exception(f"Your version of pandoc (v{pandoc_version}) is incompatible with "
```
Raise a runtime error instead of a blanket exception here
```suggestion
                    raise RuntimeError(f"Your version of pandoc (v{pandoc_version}) is incompatible with "
```

shyamd commented on src/mkdocs_bibtex/utils.py:
```diff
@@ -145,8 +214,29 @@ def insert_citation_keys(citation_quads, markdown):
+                # replacing it with whitespace added first
+                markdown = markdown.replace(f" {full_citation}", replacement_citaton)
+
+            except Exception as e:
```
What kind of errors are you trying to catch here? Catching blanket `Exception` is bad because even ctrl+c will get caught and ignored here.

sondregronas replied:
I originally intended to do a RuntimeError, but changed it print a warning and fallback to the old method instead. In hindsight just doing the RuntimeError makes more sense. Give me some minutes and I'll resolve this

shyamd reviewed (changes requested):
Looks pretty good. Few suggestions to make things a bit cleaner.

shyamd commented:
Ready to merge?

sondregronas commented:
I suppose so - this should cover all necessary features to properly support inline citations, aswell as improve parsing and extracting citekeys in general.

The only thing I'm not sure about (irrelevant to this PR) is if the regex in `extract_cite_keys` covers all edgecases for citekeys for formatting the bibliography entries, I've seen some references to citekeys looking like `[@somestring{key}]` which would work inline, but only the part up to `{` (somestring) would be extracted and written to `quads`. I presume it's still sufficient, but I've only had experience with regular zotero citekeys.

shyamd commented:
Hmm, regarding the regex issue. It's never going to be perfect. We can improve in the future :p

alexvoss commented:
Hey @shyamd and @sondregronas - thanks a lot for implementing this. I will start adding page numbers to my docs now and will report back if I find anything that does not work but should. Thanks again for the work you put in.
